File: runtime/jvm/src/lox/LoxEnumCtor.java
package lox;

/**
 * The callable constructor an `enum` declaration produces for each variant.
 * Calling it — even with zero arguments — allocates a fresh {@link LoxEnum};
 * the bare constructor name is not a value on its own (see
 * bytecode-translation-problems.md: nullary variants "do not auto-construct").
 */
public final class LoxEnumCtor implements LoxCallable {
    public final int tag;
    public final int arity;
    public final String ctorName;
    public final String enumName;

    public LoxEnumCtor(int tag, int arity, String ctorName, String enumName) {
        this.tag = tag;
        this.arity = arity;
        this.ctorName = ctorName;
        this.enumName = enumName;
    }

    @Override
    public Object call(Object[] args) {
        if (args.length != arity) {
            // Native's own text (src/vm.cpp's CALL, isEnumCtor branch) is
            // this exact literal — it does not name the constructor or
            // report either count.
            throw LoxOps.makeError("ConstructorArityError",
                    "Constructor called with wrong arity.");
        }
        return new LoxEnum(this, args.clone());
    }
}

File: runtime/clr/src/LoxEnumCtor.cs
namespace Lox;

/// <summary>
/// The callable constructor an <c>enum</c> declaration produces for each
/// variant. Calling it - even with zero arguments - allocates a fresh
/// <see cref="LoxEnum"/>; the bare constructor name is not a value on its
/// own (see bytecode-translation-problems.md: nullary variants "do not
/// auto-construct").
/// </summary>
public sealed class LoxEnumCtor : ILoxCallable {
    public readonly int Tag;
    public readonly int Arity;
    public readonly string CtorName;
    public readonly string EnumName;

    public LoxEnumCtor(int tag, int arity, string ctorName, string enumName) {
        Tag = tag;
        Arity = arity;
        CtorName = ctorName;
        EnumName = enumName;
    }

    public object Call(object[] args) {
        if (args.Length != Arity) {
            throw new LoxError($"'{CtorName}' expects {Arity} argument(s) but got {args.Length}.");
        }
        return new LoxEnum(this, (object[])args.Clone());
    }
}
